x/stakeibc/keeper/reward_converter.go

```diff
@@ -0,0 +1,442 @@
+
+	"github.com/Stride-Labs/stride/v16/utils"
+	epochstypes "github.com/Stride-Labs/stride/v16/x/epochs/types"
+	icqtypes "github.com/Stride-Labs/stride/v16/x/interchainquery/types"
+	"github.com/Stride-Labs/stride/v16/x/stakeibc/types"
+)
+
+// The goal of this code is to allow certain reward token types to be automatically traded into other types
+// This happens before the rest of the staking, allocation, distribution etc. would continue as normal
+
+// Reward tokens are any special denoms which are paid out in the withdrawal address
+// Most host zones inflate their tokens and the newly minted host denom is what appears in the withdrawal ICA
+// This code allows for chains to use foreign denoms as revenue, which can be traded to any other denom first
+
+// 0. Before the normal epochly base denom withdrawal address check in the normal staking flow
+//		by doing this check before the normal staking flow, we could trade tokens which are stake-able instead
+// 1. Epochly check the reward denom balance in the withdrawal address
+//       on callback, send all this reward denom from withdrawl ICA to trade ICA on the trade zone (OSMOSIS)
+// 2. Epochly check the reward denom balance in trade ICA
+//		on callback, trade all reward denom for target output denom defined by pool and routes in params
+// 3. Epochly check the target denom balance in trade ICA
+//		on callback, transfer these target denoms from trade ICA to withdrawal ICA on original host zone
+// Normal staking flow continues from there. So if the target denom is the original host zone base denom
+// as will often be the case, then these tokens will follow the normal staking and distribution flow.
+
+// msgs with packet forwarding memos can unwind through the reward zone and chain two transfer hops without callbacks
+
+// ICA tx will kick off transfering the reward tokens from the hostZone withdrawl ICA to the tradeZone trade ICA
+// This will be two hops to unwind the ibc denom through the rewardZone using pfm in the transfer memo if possible
+func (k Keeper) TransferRewardTokensHostToTrade(ctx sdk.Context, amount sdk.Int, route types.TradeRoute) error {
+	// Timeout for ica tx and the transfer msgs is at end of epoch
+	strideEpochTracker, found := k.GetEpochTracker(ctx, epochstypes.STRIDE_EPOCH)
+	if !found {
+		return errorsmod.Wrapf(types.ErrEpochNotFound, epochstypes.STRIDE_EPOCH)
+	}
+	timeout := uint64(strideEpochTracker.NextEpochStartTime)
+
+	startingDenom := route.RewardDenomOnHostZone
+	sendTokens := sdk.NewCoin(startingDenom, amount)
+
+	withdrawlIcaAddress := route.HostToRewardHop.FromAccount.Address
+	unwindIcaAddress := route.HostToRewardHop.ToAccount.Address
+	tradeIcaAddress := route.RewardToTradeHop.ToAccount.Address
+
+	// Import and use pfm data structures instead of this JSON if we can determine consistent module version...
+	memoJSON := fmt.Sprintf(`"forward": {"receiver": "%s","port": "%s","channel":"%s","timeout":"10m","retries": 2}`,
```

Comment:
is there an option to specify a timeout timestamp here instead of a duration?

Also, it's possible it doesn't matter if it times out after the epoch, but feels easier to reason about if we have everything time out at once. It's possible there's a race condition I'm not thinking of

Reply:
I could be wrong, but from [this code](https://github.com/cosmos/ibc-apps/blob/main/middleware/packet-forward-middleware/packetforward/ibc_middleware.go) which I believe is the most up to date version of pfm, it seems like all input values related to time are intended to be durations instead of absolute timestamps...  I agree it would be nice to enforce the epoch boundary but I don't think it should matter for this transfer because the ICQs either find the funds or they don't in which case they should find them on the next epoch

Reply:
hmm, do you think it's worth setting up the two timeouts to guarantee this succeeds or times out by the end of the epoch?

i.e. first transfer has a timeout of currentTime+3hours, and then this autopilot message has timeout duration of 3 hours

I don't want to overcomplicate things, but it does give us a bit of peace of mind if we know things don't overlap epochs

Reply:
that said, if we have good confidence that the overlap doesn't matter (because as you said the ICQ either finds it or it doesn't), then maybe it's not worth it

Reply:
From ethan on slack:
> No I would just make sure everything timeout at the epoch boundary. Most of the time it should ever even be close with 6hour stride epoch

Reply:
The only one which matters for overlap is the swap ICQ and swap ICA command with itself.  The transfers and every other ICQ/ICA can overlap without issue, if the ICQ for the reward in the trade account happens while a previous ICA swap command is actively changing that amount by swapping, then we would have a race condition where they affect each other.  Even so, it would cause the ICQ to overestimate the amount available and the swap ICA commands it calls in the callback could fail due to trying to trade more funds than exist.  But no funds are lost, there is just a 1 epoch delay when the ICA fails and then it should work fine.  To be clear, with the epoch boundary timeouts, even this issue should be impossible currently